core/node/base_token_adjuster/src/base_token_adjuster.rs

```diff
@@ -0,0 +1,216 @@
+use std::{
+    fmt::Debug,
+    time::{Duration, Instant},
+};
+
+use anyhow::Context as _;
+use async_trait::async_trait;
+use chrono::Utc;
+use rand::Rng;
+use tokio::{sync::watch, time::sleep};
+use zksync_config::configs::base_token_adjuster::BaseTokenAdjusterConfig;
+use zksync_dal::{ConnectionPool, Core, CoreDal};
+use zksync_types::{
+    base_token_price::BaseTokenAPIRatio, fee_model::BaseTokenConversionRatio, Address,
+    SHARED_BRIDGE_ETHER_TOKEN_ADDRESS,
+};
+
+#[async_trait]
+pub trait BaseTokenAdjuster: Debug + Send + Sync {
+    /// Returns the last ratio cached by the adjuster and ensure it's still usable.
+    async fn get_conversion_ratio(&self) -> anyhow::Result<BaseTokenConversionRatio>;
+}
+
+#[derive(Debug, Clone)]
+/// BaseTokenAdjuster implementation for the main node (not the External Node). TODO (PE-137): impl APIBaseTokenAdjuster
+pub struct MainNodeBaseTokenAdjuster {
+    pool: ConnectionPool<Core>,
+    base_token_l1_address: Address,
+    config: BaseTokenAdjusterConfig,
+}
+
+impl MainNodeBaseTokenAdjuster {
+    pub fn new(
+        pool: ConnectionPool<Core>,
+        config: BaseTokenAdjusterConfig,
+        base_token_l1_address: Address,
+    ) -> Self {
+        Self {
+            pool,
+            config,
+            base_token_l1_address,
+        }
+    }
+
+    /// Main loop for the base token adjuster.
+    /// Orchestrates fetching new ratio, persisting it, and updating L1.
+    pub async fn run(&mut self, stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()> {
+        let pool = self.pool.clone();
+        loop {
+            if *stop_receiver.borrow() {
+                tracing::info!("Stop signal received, base_token_adjuster is shutting down");
+                break;
+            }
+
+            let start_time = Instant::now();
+
+            match self.fetch_new_ratio().await {
+                Ok(new_ratio) => {
+                    match self.persist_ratio(&new_ratio, &pool).await {
```

Comment:
This logic is will most likely be overridden by hyper chains operator as one may want to persist the ratio only under some conditions 

 Better to move the altogether with the `run` function under `external_price_component`

Reply:
A few thoughts:
- I think this comment conflates persisting the most recent ratio with updating the L1
- I believe operators will want to customize the behavior, but not override it. They will be able to decide how often new price quotes are fetched, and under what conditions the L1 price is updated
- I also believe that operators might want to override what external APIs they're reaching out to, and so this logic is encapsulated outside of the `BaseTokenAdjuster`
- Lastly, this is injected into the node via the node framework, so operators will be able to override whatever this entire layer if any of the core logic isn't fulfilling their needs

Please lmk if you see things differently!